File: packages/axon-core-ts/src/historian.ts
/** Historian (process data) model. Used by HistorianBridge and labbench clients. */

export interface HistorianTag {
  name: string;
  description?: string;
  unit?: string;
  /** Canonical asset path, e.g., "Site/Line/Station/Sensor". */
  assetPath?: string;
}

export type Quality = "good" | "bad" | "questionable" | "uncertain";

export interface HistorianPoint {
  tag: string;
  ts: string; // ISO-8601 UTC
  value: number;
  quality?: Quality;
}

export interface TagQuery {
  tags: string[];
  from: string;
  to: string;
  /** Aggregation (default: raw). */
  agg?: "raw" | "avg" | "min" | "max" | "stddev" | "count" | "first" | "last";
  /** Sample interval for aggregations, e.g., "1m", "10s". */
  interval?: string;
}

File: packages/axon-core-py/src/axon_core/historian.py
from __future__ import annotations
from typing import Literal, Optional
from pydantic import BaseModel, Field

Quality = Literal["good", "bad", "questionable", "uncertain"]


class HistorianTag(BaseModel):
    name: str
    description: Optional[str] = None
    unit: Optional[str] = None
    assetPath: Optional[str] = None


class HistorianPoint(BaseModel):
    tag: str
    ts: str
    value: float
    quality: Optional[Quality] = None


class TagQuery(BaseModel):
    """Wire/Python TagQuery. Wire field is `from`; Python attribute is `from_`.

    Both `TagQuery(from_=...)` and `TagQuery.model_validate({"from": ...})` work.
    """

    tags: list[str]
    from_: str = Field(alias="from")
    to: str
    agg: Optional[Literal["raw", "avg", "min", "max", "stddev", "count", "first", "last"]] = "raw"
    interval: Optional[str] = None

    model_config = {"populate_by_name": True}
